### DISCUSSION TITLE
Problems with making gatsby-source-flotiq compatible with gatsby v4 when trying to build the project using it

### DISCUSSION THREAD

**Discussion by CiotkaCierpienia** [gatsbyjs/gatsby#34386] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Hi, I'm trying to make gatsby-source-flotiq compatible with v4, but I encounter a problem when building the project with it. The project works perfectly in development. The plugin does not resolve our datasources correctly (connected nodes), but infer schema from the original object, and because of that queries cannot find fields on those inferred types, the plugin: https://github.com/flotiq/gatsby-source-flotiq, the project: https://github.com/flotiq/flotiq-blog, I've unfortunately run out of ideas how to make it work. The API key from the reproduction repository will be removed in a week. The schema in cache is correct, and it makes calls to resolve functions in the schema but ignores the results when creating the pages.

This is the schema produced by the plugin:
```
...
type FlotiqBlogPost implements Node @dontInfer {
  slug: String!
  tags: [FlotiqBlogTag]
  title: String!
  author: [FlotiqBlogAuthor]
  status: String!
  content: FlotiqBlock
  excerpt: String!
  headerImage: [_media]
  publish_date: String
  metaDescription: String!
  flotiqInternal: FlotiqInternal!
}

type FlotiqBlogTag implements Node @dontInfer {
  tag: String!
  tag_name: String!
  description: String
  flotiqInternal: FlotiqInternal!
  type: String
  dataUrl: String
}
...
```

the schema that is used in the build process (as I understand it):
```
...
type FlotiqBlogPost implements Node @dontInfer {
  slug: String!
  tags: [FlotiqBlogPostTags]
  title: String!
  author: [FlotiqBlogPostAuthor]
  status: String!
  content: FlotiqBlock
  excerpt: String!
  headerImage: [FlotiqBlogPostHeader_image]
  publish_date: String
  metaDescription: String!
  flotiqInternal: FlotiqInternal!
}
type FlotiqBlogPostTags {
  type: String,
  dataUrl: String
}
...
```

FlotiqBlogPostTags  - is not created by the plugin but inferred from the original post object.

I've tried to use:
```
module.exports = {
  flags: {
    DETECT_NODE_MUTATIONS: true,
  },
}
```
as was suggested here: https://www.gatsbyjs.com/docs/how-to/local-development/debugging-missing-data/, but with no results

### Reproduction Link

https://github.com/CiotkaCierpienia/bug-repro

### Steps to Reproduce

1. Install gatsby-source-flotiq plugin
2. Use content type definition with relation in the graphql query
3. Run gatsby build

### Expected Result

The project is building correctly.

### Actual Result

```
There was an error in your GraphQL query:

Cannot query field "id" on type "FlotiqBlogPostTags".

If you don't expect "id" to exist on the type "FlotiqBlogPostTags" it is most likely a typo.
However, if you expect "id" to exist there are a couple of solutions to common problems:
```

### Environment

```shell
System:
    OS: Windows 10 10.0.18363
    CPU: (8) x64 Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz
  Binaries:
    Node: 14.17.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.14.12 - ~\AppData\Roaming\npm\npm.CMD
  Languages:
    Python: 2.7.10 - C:\Python27\python.EXE
  Browsers:
    Chrome: 96.0.4664.110
    Edge: Spartan (44.18362.449.0)
  npmPackages:
    gatsby: ^4.4.0 => 4.4.0
    gatsby-plugin-disqus: ^1.2.3 => 1.2.3
    gatsby-plugin-google-analytics: ^4.4.0 => 4.4.0
    gatsby-plugin-hotjar: ^1.1.1 => 1.1.1
    gatsby-plugin-image: ^2.4.0 => 2.4.0
    gatsby-plugin-manifest: ^4.4.0 => 4.4.0
    gatsby-plugin-react-helmet: ^5.4.0 => 5.4.0
    gatsby-plugin-robots-txt: ^1.6.14 => 1.6.14
    gatsby-plugin-sass: ^5.4.0 => 5.4.0
    gatsby-plugin-sharp: ^4.4.0 => 4.4.0
    gatsby-plugin-sitemap: ^5.4.0 => 5.4.0
    gatsby-source-filesystem: ^4.4.0 => 4.4.0
    gatsby-source-flotiq: ^4.0.0 => 4.1.0
    gatsby-transformer-sharp: ^4.4.0 => 4.4.0
  npmGlobalPackages:
    gatsby-cli: 4.3.0
```

**Comment by graysonhicks** [comment]
Hi @CiotkaCierpienia, thanks for the issue!  I got this error trying to run the reproduction locally:

```
FLOTIQ: network timeout at:
https://api.flotiq.com/api/v1/internal/contenttype?limit=10000&order_by=label
```

I'm guessing its related to the API key?

**Comment by CiotkaCierpienia** [comment]
Yes, it is related to the API key, we didn't have timeout issues for a long time. I could fetch everything locally in the reproduction repo, but to be sure that it will work for you also, I extended the timeout for that repo.

**Comment by LekoArts** [comment]
Hi,

please see our migration guides: https://www.gatsbyjs.com/docs/reference/release-notes/migrating-from-v3-to-v4/ & https://www.gatsbyjs.com/docs/reference/release-notes/migrating-source-plugin-from-v3-to-v4/

The plugin is in it's current state not yet compatible with Gatsby 4, e.g. you're still creating type definitions in `sourceNodes` which is not allowed anymore:

https://github.com/flotiq/gatsby-source-flotiq/blob/a17708616c948e3490a25403ed93a1f5680c3851/gatsby-node.js#L68

You'll also need to move the image downloading out of `createResolvers`: https://github.com/flotiq/gatsby-source-flotiq/blob/a17708616c948e3490a25403ed93a1f5680c3851/gatsby-node.js#L198
